### DISCUSSION TITLE
Is it possible to type hint `tuple` and `dict` equivalent to `P.args` and `P.kwargs`?

### DISCUSSION THREAD

**Discussion by randolf-scholz** [python/typing#1447] [Q&A] [answered]
Is it possible to "unpack" `ParamSpec` args and kwargs into a `tuple`/`dict`?

Context: let's annotate a function that takes a `Callable`, it's `args` and `kwargs` and evaluates the callable. This can be type hinted successfully as:

```python
def eval_func(func: Callable[P, R], /, *args: P.args, **kwargs: P.kwargs) -> R:
    return func(*args, **kwargs)
```

Is it possible to properly type-hint an equivalent function that takes as input a single triplet?

```python
def eval_triplet(triplet: tuple[Callable[..., R], tuple, dict], /) -> R:
    func, args, kwargs = triplet
    return func(*args, **kwargs)
```

What I mean is if we replace `...` with `P` in the above signature, is it possible to replace `tuple` and `dict` so that we get the same detailed type-checking for `eval_triplet` as with `eval_func`? (intuitively a syntax like `tuple[*P.args]` and `dict[**P.kwargs]` comes to mind, but this is not supported)

Full example: ([mypy-playground](https://mypy-play.net/?mypy=latest&python=3.11&gist=e3b59a56841733b2715c826f0427b9b3))

```python
from typing import Callable, ParamSpec, TypeVar

P = ParamSpec("P")
R = TypeVar("R")

def eval_func(func: Callable[P, R], /, *args: P.args, **kwargs: P.kwargs) -> R:
    return func(*args, **kwargs)

def eval_triplet(triplet: tuple[Callable[..., R], tuple, dict], /) -> R:
    func, args, kwargs = triplet
    return func(*args, **kwargs)

def mandatory_positional(x: int, y: int, /) -> int:
    return x+y

# calling with correct signature
eval_func(mandatory_positional, 1, 2)  # ✔
eval_triplet((mandatory_positional, (1,2), {}))  # ✔

# calling with completely wrong signature
eval_func(mandatory_positional)   # ✔ correctly raises
eval_triplet((mandatory_positional, (), {}))  # ✘ does not raise...
```

**Comment by erictraut** [comment, marked as answer]
No, there is no mechanism currently in the type system that enables this.

You can use unpack operators to make this work with the available mechanism, although some errors may go uncaught by current type checker implementations because this is not a typical usage pattern.

```python
eval_func(mandatory_positional, *(1, 2), **{}) # No error (correct)
eval_func(mandatory_positional, *("1", 2), **{}) # Error in both mypy and pyright (correct)
eval_func(mandatory_positional, *(1, 2, 3), **{}) # Error in mypy, false negative in pyright
eval_func(mandatory_positional, *(), **{"a": 1, "b": 2})  # False negative in both mypy and pyright
```
